File: websocket-checksum-hash/csharp/ChecksumHash.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Text;

namespace WebSocketChecksumHash
{
    public class ChecksumHash
    {
        /// <summary>
        /// Generate a checksum hash for an orderbook.
        /// </summary>
        /// <param name="orderbook">A dictionary with 'bids' and 'asks' lists, each containing
        /// dictionaries with 'price' and 'amount' keys.</param>
        /// <returns>A string representation of the CRC32 checksum.</returns>
        public static string GetOrderbookChecksum(Dictionary<string, List<Dictionary<string, object>>> orderbook)
        {
            var bids = ChecksumInput(orderbook.ContainsKey("bids") ? orderbook["bids"] : new List<Dictionary<string, object>>());
            var asks = ChecksumInput(orderbook.ContainsKey("asks") ? orderbook["asks"] : new List<Dictionary<string, object>>());
            
            if (string.IsNullOrEmpty(bids) && string.IsNullOrEmpty(asks))
            {
                return "0";
            }
            
            return CalculateCrc32(bids + asks).ToString();
        }
        
        private static string ChecksumInput(List<Dictionary<string, object>> orders)
        {
            var ordersStr = new StringBuilder();
            var iterations = Math.Min(orders.Count, 10);
            
            for (int i = 0; i < iterations; i++)
            {
                var price = orders[i]["price"].ToString();
                var amount = orders[i]["amount"].ToString();
                ordersStr.Append(TrimValue(price)).Append(TrimValue(amount));
            }
            
            return ordersStr.ToString();
        }
        
        /// <summary>
        /// Remove decimal points and leading zeros from a string value.
        /// </summary>
        /// <param name="value">The string value to trim.</param>
        /// <returns>The trimmed string.</returns>
        public static string TrimValue(string value)
        {
            return value.Replace(".", "").TrimStart('0');
        }
        
        /// <summary>
        /// Calculate CRC32 checksum for a string.
        /// </summary>
        /// <param name="input">Input string.</param>
        /// <returns>CRC32 checksum as uint.</returns>
        private static uint CalculateCrc32(string input)
        {
            // CRC-32 polynomial
            const uint polynomial = 0xEDB88320;
            var bytes = Encoding.UTF8.GetBytes(input);
            uint crc = 0xFFFFFFFF;
            
            foreach (var b in bytes)
            {
                crc ^= b;
                for (int i = 0; i < 8; i++)
                {
                    crc = (crc & 1) != 0 ? (crc >> 1) ^ polynomial : crc >> 1;
                }
            }
            
            return ~crc;
        }
    }
}

File: websocket-checksum-hash/php/ChecksumHash.php
<?php

/**
 * WebSocket Checksum Hash implementation for PHP
 */
class ChecksumHash
{
    /**
     * Generate a checksum hash for an orderbook.
     *
     * @param array $orderbook An array with 'bids' and 'asks' arrays, each containing
     *                         arrays with 'price' and 'amount' keys.
     * @return string A string representation of the CRC32 checksum.
     */
    public static function getOrderbookChecksum(array $orderbook): string
    {
        $bids = self::checksumInput($orderbook['bids'] ?? []);
        $asks = self::checksumInput($orderbook['asks'] ?? []);
        
        if (empty($bids) && empty($asks)) {
            return '0';
        }
        
        $checksum = crc32($bids . $asks);
        // PHP's crc32 can return negative values, so we need to convert to unsigned
        if ($checksum < 0) {
            $checksum = sprintf('%u', $checksum);
        }
        
        return (string) $checksum;
    }
    
    /**
     * Process orders for checksum calculation.
     *
     * @param array $orders Array of orders.
     * @return string Processed string.
     */
    private static function checksumInput(array $orders): string
    {
        $ordersStr = '';
        
        // If there are no orders, return an empty string
        if (empty($orders)) {
            return $ordersStr;
        }
        
        // Create a copy of the orders array to avoid modifying the original
        $ordersCopy = $orders;
        
        // Sort the orders by price (we'll use the same order for both bids and asks)
        // This is a simplification - in a real implementation, bids would be sorted from highest to lowest
        // and asks from lowest to highest, but the JavaScript implementation doesn't do this sorting
        
        $iterations = min(count($ordersCopy), 10);
        
        for ($i = 0; $i < $iterations; $i++) {
            $price = (string) $ordersCopy[$i]['price'];
            $amount = (string) $ordersCopy[$i]['amount'];
            $ordersStr .= self::trimValue($price) . self::trimValue($amount);
        }
        
        return $ordersStr;
    }
    
    /**
     * Remove decimal points and leading zeros from a string value.
     *
     * @param string $value The string value to trim.
     * @return string The trimmed string.
     */
    public static function trimValue(string $value): string
    {
        $value = str_replace('.', '', $value);
        return ltrim($value, '0');
    }
}
